R/format.R

```diff
@@ -90,11 +143,139 @@ xportr_format <- function(.df,
 
   names(format) <- filtered_metadata[[variable_name]]
 
+  # vector of expected formats for clinical trials (usually character or date/time)
+  # https://documentation.sas.com/doc/en/pgmsascdc/9.4_3.5/leforinforref
+  # /n0p2fmevfgj470n17h4k9f27qjag.htm#n0wi06aq4kydlxn1uqc0p6eygu75
+
+  expected_formats <- c(
+    NA,
+    "",
+    paste("$", 1:200, ".", sep = ""),
+    paste("date", 5:11, ".", sep = ""),
+    paste("time", 2:20, ".", sep = ""),
+    paste("datetime", 7:40, ".", sep = ""),
+    paste("yymmdd", 2:10, ".", sep = ""),
+    paste("mmddyy", 2:10, ".", sep = ""),
+    paste("ddmmyy", 2:10, ".", sep = ""),
+    "E8601DA.",
+    "E8601DA10.",
+    "E8601DN.",
+    "E8601DN10.",
+    "E8601TM.",
+    paste("E8601TM", 8:15, ".", sep = ""),
+    paste("E8601TM", 8:15, ".", sort(rep(0:6, 8)), sep = ""),
+    "E8601TZ.",
+    paste("E8601TZ", 9:20, ".", sep = ""),
+    paste("E8601TZ", 9:20, ".", sort(rep(0:6, 12)), sep = ""),
+    "E8601TX.",
+    paste("E8601TX", 9:20, ".", sep = ""),
+    "E8601DT.",
+    paste("E8601DT", 16:26, ".", sep = ""),
+    paste("E8601DT", 16:26, ".", sort(rep(0:6, 11)), sep = ""),
+    "E8601LX.",
+    paste("E8601LX", 20:35, ".", sep = ""),
+    "E8601LZ.",
+    paste("E8601LZ", 9:20, ".", sep = ""),
+    "E8601DX.",
+    paste("E8601DX", 20:35, ".", sep = ""),
+    "B8601DT.",
+    paste("B8601DT", 15:26, ".", sep = ""),
+    paste("B8601DT", 15:26, ".", sort(rep(0:6, 12)), sep = ""),
+    "IS8601DA.",
+    "B8601DA.",
+    paste("B8601DA", 8:10, ".", sep = ""),
+    "weekdate.",
+    paste("weekdate", 3:37, ".", sep = ""),
+    "mmddyy.",
+    "ddmmyy.",
+    "yymmdd.",
+    "date.",
+    "time.",
+    "hhmm.",
+    "IS8601TM.",
+    "E8601TM.",
+    "B8601TM."
+  )
+
+  # w.d format for numeric variables
+  format_regex <- "^([1-9]|[12][0-9]|3[0-2])\\.$|^([1-9]|[12][0-9]|3[0-2])\\.([1-9]|[12][0-9]|3[0-1])$"
+
+
   for (i in seq_len(ncol(.df))) {
     format_sas <- purrr::pluck(format, colnames(.df)[i])
     if (is.na(format_sas) || is.null(format_sas)) {
       format_sas <- ""
     }
+    # series of checks for formats
+
+    # check that any variables ending DT, DTM, TM have a format
+    if (grepl("DT$|DTM$|TM$", colnames(.df)[i]) == TRUE && format_sas == "") {
+      message <- glue(
+        "(xportr::xportr_format) {encode_vars(colnames(.df)[i])} is expected to have a format but does not."
+      )
+      xportr_logger(message, type = "warn")
+    }
+
+    # remaining checks to be carried out if a format exists
+    if (format_sas != "") {
+      # if the variable is character
+      if (class(.df[[i]])[1] == "character") {
```

Comment:
Is there a way to use some `else if` statement in this logic since the branching is so similar.  @averissimo @vedhav @elimillera  do you all see a way to reduce the complexity here.

Reply:
Will have a look into this... but if anyone else has any suggestions, please let me know - it would be much appreciated!

Reply:
We can just pass this issue to next release.  Can you make an issue for this one?

I think it is more important to focus on tests

Reply:
Issue created! https://github.com/atorus-research/xportr/issues/251